freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/SudoRuleService.java

```diff
@@ -0,0 +1,163 @@
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Service;
+
+import com.google.common.base.Strings;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupRequest;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupService;
+import com.sequenceiq.freeipa.client.FreeIpaClient;
+import com.sequenceiq.freeipa.client.FreeIpaClientException;
+import com.sequenceiq.freeipa.client.RetryableFreeIpaClientException;
+import com.sequenceiq.freeipa.client.model.SudoCommand;
+import com.sequenceiq.freeipa.client.model.SudoRule;
+import com.sequenceiq.freeipa.entity.Stack;
+
+@Service
+public class SudoRuleService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SudoRuleService.class);
+
+    private static final String HOST_CATEGORY = "all";
+
+    @Value("${freeipa.sudo.ruleName}")
+    private String ruleName;
+
+    @Value("${freeipa.sudo.allowCommands:}")
+    private Set<String> allowCommands;
+
+    @Value("${freeipa.sudo.denyCommands:}")
+    private Set<String> denyCommands;
+
+    private Set<String> allCommands;
+
+    @Inject
+    private VirtualGroupService virtualGroupService;
+
+    @PostConstruct
+    public void postConstruct() {
+        allCommands = Stream.concat(allowCommands.stream(), denyCommands.stream()).collect(Collectors.toSet());
+    }
+
+    @Retryable(value = RetryableFreeIpaClientException.class,
+            maxAttemptsExpression = RetryableFreeIpaClientException.MAX_RETRIES_EXPRESSION,
+            backoff = @Backoff(delayExpression = RetryableFreeIpaClientException.DELAY_EXPRESSION,
+                    multiplierExpression = RetryableFreeIpaClientException.MULTIPLIER_EXPRESSION))
+    public void setupSudoRule(Stack stack, FreeIpaClient freeIpaClient) throws Exception {
+        if (stack.isAvailable()) {
+            String group = getSudoGroupName(stack.getEnvironmentCrn());
+            if (!Strings.isNullOrEmpty(group)) {
+                validateGroupExistsOnFreeIpa(group, freeIpaClient);
+
+                LOGGER.info("Setup sudo rule '{}' to allow '{}' but deny '{}' commands for user group '{}' on all hosts.",
+                        ruleName, allowCommands, denyCommands, group);
+
+                SudoRule sudoRule = getOrCreateSudoRule(freeIpaClient);
+                if (HOST_CATEGORY.equals(sudoRule.getHostCategory())) {
+                    createMissingSudoCommands(sudoRule, freeIpaClient);
+                    addMissingSudoRuleAllowCommands(sudoRule, freeIpaClient);
+                    addMissingSudoRuleDenyCommands(sudoRule, freeIpaClient);
+                    addMissingSudoUserGroup(sudoRule, group, freeIpaClient);
+                    LOGGER.info("Sudo rule '{}' setup successfully finished.", ruleName);
+                } else {
+                    throw new IllegalStateException("Failed to setup sudo rule as '" + ruleName + "' is already exists but the host category is not 'all'.");
```

Comment:
this would fail the whole usersync, is this expeted? usually if something goes wrong during usersync we just add tthe failure to a warning list

Reply:
Yes, you are right. I'm going to wrap the whole service implementation with a big try and add all exceptions to the warnings...

Reply:
@lacikaaa Because of the previously requested Retryable mechanism, we should have the surrounding try catch at the caller side. To be more precise, in UserSyncForStackService. Can you accept it?

Reply:
or maybe we can have new method in SudoRuleService what calls the retyiable one, catches all errors and add them to warnings... what do you more prefer?

Reply:
- on the caller side is not ok as other methods would be retried too
- calling inside the class won't work, you have to create another class then

Reply:
as we have discussed, it is fixed by a caller side try-catch